src/AmalgamatorSession.ts

```diff
@@ -721,12 +721,14 @@ export class AmalgamatorSession extends LoggingDebugSession {
                     const threadInfo = threadOfChildDap as ThreadInfo;
                     if (
                         threadInfo.running === false &&
+                        thread.name.includes(threadInfo.name) &&
```

Comment:
I don't understand this. But my guess is the logic is related to iterating through all the threads multiple times and some of the children having multiple threads? 

The loop at 714 `for (const thread of threads) {` iterates through all threads, and then loop at 720 `for (const threadOfChildDap of childResponse.body.threads) {` loops through all theads again. 

Is it possible that the nested loops are iterating sort of like this:

Line 713 threads contains: `RH850 1`, `RH850 2`, `RH850 3`, `RH850 4`, `RL 1`
Line 718 will then extract child dap for each thread, meaning line 719 will request threads for RH850 3 different time and RL once.
That means at line 721 threadInfo will be something like the following on each iteration: `RH850 1`, `RH850 2`, `RH850 3`, `RH850 4`, `RH850 1`, `RH850 2`, `RH850 3`, `RH850 4`, `RH850 1`, `RH850 2`, `RH850 3`, `RH850 4`, `RL 1`

That means that all the RH850 are being processed too often.

Please let me know if that makes sense and if that is what you are observing. If it is, please fix the two loops so that you are iterating through all the threads only once in the first place.

e.g. does this work to replace lines 719 onwards:

```typescript
if (thread.running === false && command === 'cdt-amalgamator/resumeAll') {
    await childDap.continueRequest({ threadId: childId });
    this.sendEvent(new ContinuedEvent(thread.id, false));
} else if (thread.running === true && command === 'cdt-amalgamator/suspendAll') {
   /// put code here
}
```

And also store the running state in `collectChildTheads` and change the type to `ThreadInfo` to keep the running state passed around. (I haven't tested this code, so it may not be quite right)

```diff
diff --git a/src/AmalgamatorSession.ts b/src/AmalgamatorSession.ts
index 5bb7055..ad24956 100644
--- a/src/AmalgamatorSession.ts
+++ b/src/AmalgamatorSession.ts
@@ -380,7 +380,7 @@ export class AmalgamatorSession extends LoggingDebugSession {
         }
     }
     private threadMapInProcess:
-        | Promise<[Map<number, [number, number]>, DebugProtocol.Thread[]]>
+        | Promise<[Map<number, [number, number]>, ThreadInfo[]]>
         | undefined;
     protected async getThreadMap(): Promise<Map<number, [number, number]>> {
         return new Promise<Map<number, [number, number]>>(
@@ -392,7 +392,7 @@ export class AmalgamatorSession extends LoggingDebugSession {
         );
     }
     protected getThreadMapInternal(): Promise<
-        [Map<number, [number, number]>, DebugProtocol.Thread[]]
+        [Map<number, [number, number]>, ThreadInfo[]]
     > {
         if (this.threadMapInProcess === undefined) {
             return this.collectChildTheads();
@@ -401,10 +401,10 @@ export class AmalgamatorSession extends LoggingDebugSession {
     }
 
     private collectChildTheads(): Promise<
-        [Map<number, [number, number]>, DebugProtocol.Thread[]]
+        [Map<number, [number, number]>, ThreadInfo[]]
     > {
         this.threadMapInProcess = new Promise((resolve, _reject) => {
-            const threads: DebugProtocol.Thread[] = [];
+            const threads: ThreadInfo[] = [];
             Promise.all(this.childDaps.map((dc) => dc.threadsRequest())).then(
                 (responses) => {
                     const threadMap: Map<number, [number, number]> = new Map<
@@ -421,7 +421,11 @@ export class AmalgamatorSession extends LoggingDebugSession {
                                         ? this.childDapNames[i] + ': '
                                         : ''
                                 } ${t.name}`, // XXX: prefix name here with which child this came from? What about the id of the child?
-                            } as DebugProtocol.Thread);
+                                running:
+                                    'running' in t
+                                        ? (t as ThreadInfo).running
+                                        : undefined,
+                            } as ThreadInfo);
                             threadMap.set(clientId, [i, t.id]);
                             clientId++;
                         });

```

Reply:
> Please let me know if that makes sense and if that is what you are observing. If it is, please fix the two loops so that you are iterating through all the threads only once in the first place.

I have debugged and found that to be true for the part you showed above. 
I also changed some code based on your suggestion, with this change it worked fine.
Could you please help me review it again?

Also if this code is merged, can you please build the latest version? we need it to test for an upcoming release.

Thank you so much!